WIP:  DRBG: implement get_nonce() callback

mspncp opened the pull request:
In commit 2a70d65b99e1 (pr #5503) a fallback was added which adds a random nonce
of security_strength/2 bits if no nonce callback is provided. This raises the
entropy requirements form 256 to 384 bit, which can cause problems on some
platforms (e.g. VMS, see issue #5849).

The requirements for the nonce are given in section 8.6.7 of NIST SP 800-90Ar1:

  A nonce may be required in the construction of a seed during instantiation
  in order to provide a security cushion to block certain attacks.
  The nonce shall be either:

  a. A value with at least (security_strength/2) bits of entropy, or

  b. A value that is expected to repeat no more often than a
     (security_strength/2)-bit random string would be expected to repeat.

  Each nonce shall be unique to the cryptographic module in which instantiation
  is performed, but need not be secret. When used, the nonce shall be considered
  to be a critical security parameter.

This commit implements a nonce of type b in order to lower the entropy
requirements during instantiation back to 256 bits.

The formulation "shall be unique to the cryptographic module" above implies
that the nonce needs to be unique among (with high probability) among all
DRBG instances in "space" and "time". We try to achieve this goal by creating a
nonce of the following form

    nonce = app-specific-data || high-resolution-utc-timestamp || counter

Where || denotes concatenation. The application specific data can be something
like the process or group id of the application. A utc timestamp is used because
it increases monotonically, provided the system time is synchronized.

This approach may not be perfect yet for a FIPS evaluation, but it should be good
enough for the moment.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

mspncp commented:
@levitte, I have some TODO's for you: [here](https://github.com/openssl/openssl/pull/5920/files#diff-b6c54ea0dc5b1d747760bb57b1b90d85R158) and [here](https://github.com/openssl/openssl/pull/5920/files#diff-b6c54ea0dc5b1d747760bb57b1b90d85R197).  See also the [documenting comments](https://github.com/openssl/openssl/pull/5920/files#diff-08e01c09747c1a07a7b80c69006b8bd1R87) for the prototypes.

mspncp commented on crypto/include/internal/rand_int.h:
```diff
@@ -39,8 +39,15 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
                                unsigned char *out, size_t outlen);
+size_t rand_drbg_get_nonce(RAND_DRBG *drbg,
+                           unsigned char **pout,
+                           int entropy, size_t min_len, size_t max_len);
```
copy & paste error :wink:

mspncp replied:
The `min_len` is also not necessary. In that point I will follow Kurt, see #5909.

mspncp replied:
Oh, wait! The  `entropy` argument is there since FIPS 2.0. I will have to check that first...

```C
    int FIPS_drbg_set_callbacks(DRBG_CTX * dctx,
                                size_t (*get_entropy) (DRBG_CTX * ctx,
                                                       unsigned char **pout,
                                                       int entropy,
                                                       size_t min_len,
                                                       size_t max_len),
                                void (*cleanup_entropy) (DRBG_CTX * ctx,
                                                         unsigned char *out,
                                                         size_t olen),
                                size_t entropy_blocklen,
                                size_t (*get_nonce) (DRBG_CTX * ctx,
                                                     unsigned char **pout,
                                                     int entropy,
                                                     size_t min_len,
                                                     size_t max_len),
                                void (*cleanup_nonce) (DRBG_CTX * ctx,
                                                       unsigned char *out,
                                                       size_t olen));
```

mspncp replied:
The parameter makes sense for variant a. And it makes also sense to change the name to `security_strength` as Kurt did [here](https://github.com/openssl/openssl/pull/5909/files#diff-2fea19f1443a1415d4686e753ac201d1R40).

levitte replied:
It makes sense for variant a, sure...  but this PR is to implement variant b, is it not?

mspncp replied:
For us the parameter is not necessary. But if you view the `RAND_DRBG` "class" as a general purpose building block and not just how it is wired into our RAND_METHOD, then the parameter _does_ make sense. The callbacks (which are present since FIPS 2.0) give the developer full control over the reseeding process. And if the developer provides a get_nonce() callback he needs to know the security level.

That we have a fallback for missing get_entropy callbacks is the consequence of a misunderstanding (we thought they were optional) and a convenience: Retrieving entropy + random_nonce in a single call is  more efficient than to make two generate calls, one in get_entropy() and one in get_nonce().

levitte replied:
Okie

levitte commented:
Ok, I can't quite follow the code, so I need an answer.  Are the data from `rand_pool_add_additional_data` and `rand_pool_add_nonce_data` ever combined?  If that is the case, I'm not sure I understand why the same data should ever appear in both.  For example, the Unix implementation has the result of `get_time64` in both.  Granted, there might be a difference of a milisecond if we're lucky, but...

levitte commented:
I guess that I'm a bit confused about what's appropriate to return as a nonce and what's appropriate to return as additional data...

levitte commented on crypto/rand/rand_unix.c:
```diff
@@ -256,5 +358,108 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+
+    /*
+     * Add gid, pid, and uid to nonce to ensure variation between
+     * different processes.
```
Would it be wise to have the thread ID here as well?  On some systems, that's the only way to differentiate between threads...
(as a matter of fact, wouldn't pid and tid be enough?)

mspncp commented:
> Ok, I can't quite follow the code, so I need an answer. Are the data from rand_pool_add_additional_data and rand_pool_add_nonce_data ever combined?

The nonce is used only once for instantiation, while the additional data is used on every call to `RAND_bytes()` or `RAND_priv_bytes()`, and for reseeding. (Both functions call `RAND_DRBG_bytes()` which adds the additional data to the `RAND_DRBG_generate()` call, see #4752).

For the nonce, the emphasis is on uniqueness, while for the additional data, it's more about adding a little noise and about speed (because it's called much more often).

Two examples:

- Adding the pid to the nonce improves uniqueness, but adding it as additional data on every generate call does not add any noise.

- In rand_win.c, I call `GetSystemTimeAsFileTime()` in `rand_pool_add_nonce_data()` to get a full sized monotonic UTC-Time, while in `rand_pool_add_additional_data()` I use the simpler (and faster) `RAND_rdtsc()` (falling back to `QueryPerformanceCounter()` on error), although `OPENSSL_rdtsc` only returns an `uint32_t` (there was an endless discussion about the return value in #4757).

Hope that helps? If not, let me know.

levitte commented:
Yes, that helps, thank you.

mspncp commented:
> I guess that I'm a bit confused about what's appropriate to return as a nonce and what's appropriate to return as additional data...

It's ok if you give me some code samples for obtaining the data, I can move them around as needed. I'm still figuring out for myself, what is best. Of course, suggestions are welcome.

levitte commented:
See #5922 😉

mspncp commented:
So I will pick the cherries and cut down the tree¹, right?  ;-)

¹ i.e., close #5922

mspncp commented:
Cherry-picked e35c49b55d79 from #5922, and rebased, squashing the fixups.

kroeckx commented on crypto/rand/rand_unix.c:
```diff
@@ -14,6 +14,122 @@
+#endif
+
+/* Macro to convert two thirty two bit values into a sixty four bit one */
+# define TWO32TO64(a, b) ((((uint64_t)(a)) << 32) + (b))
```
This seems a duplicate of line 61

kroeckx replied:
(And the next 30 or so line too)

mspncp replied:
All good things come in threes... :wink:  I probably copied the block in order to reindent the preprocessor commands.

Removed, thanks!

levitte commented:
I ran this on VMS as well.  Works without issues.

levitte commented:
Is this still WIP?

mspncp commented:
I want to review this myself once more and then squash it and take it out of WIP. This will happen maybe this evening.

mspncp commented:
> Would it be wise to have the thread ID here as well? On some systems, that's the only way to differentiate between threads...
(as a matter of fact, wouldn't pid and tid be enough?)

I decided to unify all callbacks to return the following data

- rand_pool_nonce_data:  process id, thread id, and high resolution (current) timestamp
- rand_pool_add_additional_data:  some noise from a high resolution timer 

Rebased and squashed the pull request. Ready for review.

mspncp commented:
Added the thread id back in for the additional data, because it _does_ add some noise if the drbg is accessed concurrently (which is the case for master).

Also added `[extended tests]`.

levitte commented:
From Travis:

```
../test/recipes/04-test_asn1_encode.t .............. 
1..1
    # Subtest: ../../test/asn1_encode_test
    1..6
    ok 1 - test_long_32bit
==15501==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7f5cc16826d0 in AES_encrypt /home/travis/build/openssl/openssl/crypto/aes/aes_core.c:888:13
    #1 0x7f5cc1860db3 in aes_ecb_cipher /home/travis/build/openssl/openssl/crypto/evp/e_aes.c:2681:9
    #2 0x7f5cc1882d8d in EVP_EncryptUpdate /home/travis/build/openssl/openssl/crypto/evp/evp_enc.c:332:13
    #3 0x7f5cc1923deb in ctr_BCC_block /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:76:10
    #4 0x7f5cc1923deb in ctr_BCC_blocks /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:88
    #5 0x7f5cc1923deb in ctr_BCC_update /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:133
    #6 0x7f5cc1923deb in ctr_df /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:197
    #7 0x7f5cc1923deb in ctr_update /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:264
    #8 0x7f5cc191f316 in drbg_ctr_instantiate /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:293:10
    #9 0x7f5cc1928ec2 in RAND_DRBG_instantiate /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:334:10
    #10 0x7f5cc192baa0 in drbg_setup /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:872:5
    #11 0x7f5cc192baa0 in do_rand_drbg_init /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:899
    #12 0x7f5cc192baa0 in do_rand_drbg_init_ossl_ /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:886
    #13 0x7f5cc11f3a7f in __GI___pthread_once (/lib/x86_64-linux-gnu/libpthread.so.0+0xda7f)
    #14 0x7f5cc19a209a in CRYPTO_THREAD_run_once /home/travis/build/openssl/openssl/crypto/threads_pthread.c:113:9
    #15 0x7f5cc192bd09 in RAND_DRBG_get0_public /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:1018:10
    #16 0x7f5cc192cb97 in drbg_bytes /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:935:23
    #17 0x7f5cc193074b in RAND_bytes /home/travis/build/openssl/openssl/crypto/rand/rand_lib.c:742:16
    #18 0x4a14a6 in do_print_item /home/travis/build/openssl/openssl/test/asn1_encode_test.c:720:11
    #19 0x4a14a6 in test_intern /home/travis/build/openssl/openssl/test/asn1_encode_test.c:818
    #20 0x4a37fb in run_tests /home/travis/build/openssl/openssl/test/testutil/driver.c:205:19
    #21 0x4a43ad in main /home/travis/build/openssl/openssl/test/testutil/main.c:51:15
    #22 0x7f5cc071af44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #23 0x41b63e in _start (/home/travis/build/openssl/openssl/test/asn1_encode_test+0x41b63e)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/travis/build/openssl/openssl/crypto/aes/aes_core.c:888:13 in AES_encrypt
Exiting
../../util/shlib_wrap.sh ../../test/asn1_encode_test => 77
not ok 1 - running asn1_encode_test

#   Failed test 'running asn1_encode_test'
#   at /home/travis/build/openssl/openssl/test/../util/perl/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
```

levitte commented:
I've re-verified that this works on VMS

mspncp commented:
Travis failures are related and all of the following type:

```
==18471==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7f3562dba6d0 in AES_encrypt /home/travis/build/openssl/openssl/crypto/aes/aes_core.c:888:13
    #1 0x7f3562f98db3 in aes_ecb_cipher /home/travis/build/openssl/openssl/crypto/evp/e_aes.c:2681:9
    #2 0x7f3562fbad8d in EVP_EncryptUpdate /home/travis/build/openssl/openssl/crypto/evp/evp_enc.c:332:13
    #3 0x7f356305bdeb in ctr_BCC_block /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:76:10
    #4 0x7f356305bdeb in ctr_BCC_blocks /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:88
    #5 0x7f356305bdeb in ctr_BCC_update /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:133
    #6 0x7f356305bdeb in ctr_df /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:197
    #7 0x7f356305bdeb in ctr_update /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:264
    #8 0x7f3563057316 in drbg_ctr_instantiate /home/travis/build/openssl/openssl/crypto/rand/drbg_ctr.c:293:10
    #9 0x7f3563060ec2 in RAND_DRBG_instantiate /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:334:10
    #10 0x7f3563063aa0 in drbg_setup /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:872:5
    #11 0x7f3563063aa0 in do_rand_drbg_init /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:899
    #12 0x7f3563063aa0 in do_rand_drbg_init_ossl_ /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:886
    #13 0x7f356292ba7f in __GI___pthread_once (/lib/x86_64-linux-gnu/libpthread.so.0+0xda7f)
    #14 0x7f35630da09a in CRYPTO_THREAD_run_once /home/travis/build/openssl/openssl/crypto/threads_pthread.c:113:9
    #15 0x7f35630642d9 in RAND_DRBG_get0_private /home/travis/build/openssl/openssl/crypto/rand/drbg_lib.c:1038:10
    ...
```

mspncp commented:
Oh, I didn''t see your post. Was occupied with explaining some math to my son. :wink:

mspncp commented:
So the the uninitialized memory issue was caused by an alignment gap (between pid and tid) in the `data` struct. I could have fixed it by reordering the entries or by adding a `#pragma pack`, but this would have added unnecessary complexity. So I just zeroed the `data`before filling it.

```
(gdb) frame
#0  rand_pool_add_nonce_data (pool=0x5555557c8f70) at crypto/rand/rand_unix.c:304
304	    data.pid = getpid();
(gdb) p data
$6 = {
  pid = 0, 
  tid = 0, 
  time = 0
}
(gdb) p sizeof(data)
$7 = 24
(gdb) p sizeof(data.pid)
$8 = 4
(gdb) p sizeof(data.tid)
$9 = 8
(gdb) p sizeof(data.time)
$10 = 8
```

mspncp commented:
Oh, and even the coveralls test is happy about an extra beer (coverage increased by 0.04 ‰)

mspncp commented:
Thanks Richard! I will not merge before Friday evening, leaving others the opportunity to comment.

mspncp commented:
I'll add a "Fixes #5849" when I merge.

kroeckx commented on crypto/rand/rand_unix.c:
```diff
@@ -256,5 +285,113 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+     */
+    data.pid = getpid();
+    data.tid = CRYPTO_THREAD_get_current_id();
+    data.time = get_time64();
```
This would be good for additional data, not the nonce.

mspncp replied:
I totally agree with you. The perfect nonce ingredient would be a 64 bit high resolution utc timestamp (like it's the case for Windows), because that decreases the overall collision probability drastically, provided the computer's clocks are synchronized properly. Unfortunately, it seems to be almost impossible to get that for all unix-ish platforms in a portable way. I'm sure you remember the struggle we had with the timers in #5231. I'm also not happy with the `#ifdef` spaghetti in `get_time64()`, but it seems we had no other choice. So if you have any suggestions how to improve that, I would be happy to hear about it.

mspncp replied:
One could use different timers for nonce and additional data, the former optimized for uniqueness and the latter for speed, Since the noise is in the lower bits, OPENSSL_rdtsc() would be sufficient for additional data.

kroeckx replied:
There are things like gettimeofday() and time(), which I think should be supported on all POSIX platforms.

kroeckx replied:
Note that I don't see 64 bit as a requirement, gettimeofday() should be good enough.

mspncp commented on crypto/rand/rand_unix.c:
```diff
@@ -256,5 +285,113 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+        cid = CLOCK_REALTIME;
+#  endif
+
+        if (clock_gettime(cid, &ts) == 0)
```
So maybe we should first attempt a `clock_gettime()` with `cid=CLOCK_REALTIME` before trying anything else, right?

mspncp commented:
I split the `get_time64()` function into two functions, one optimized for the nonce, the other one optimized for additional data.

mspncp commented on crypto/rand/rand_unix.c:
```diff
@@ -348,50 +381,34 @@ static uint64_t get_time64(void)
             return TWO32TO64(tv.tv_sec, tv.tv_usec);
     }
+#  endif
+    return time(NULL);
```
The manual page says that time() never fails if a NULL pointer is passed.

mspncp commented on crypto/rand/rand_unix.c:
```diff
@@ -348,50 +381,34 @@ static uint64_t get_time64(void)
     {
         struct timeval tv;
 
         if (gettimeofday(&tv, NULL) == 0)
```
The manual page says that the call only fails if the args are out of range, which is not the case here. Is it really necessary to check the return value?

mspncp commented on crypto/rand/rand_unix.c:
```diff
@@ -348,50 +381,34 @@ static uint64_t get_time64(void)
+#   endif
 
-        if (clock_gettime(cid, &ts) == 0)
+        if (clock_gettime(CLOCK_TYPE, &ts) == 0)
```
BTW: The manual page says that the call only fails if the pointer is out of range or the clock type is not supported. So this call should never fail. Is it really necessary to check the return value?

mspncp commented:
Kurt, I would like to merge this evening if possible. If you need more time to review, please let me know.

kroeckx commented on crypto/rand/rand_unix.c:
```diff
@@ -69,17 +110,6 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
-    curr_pid = getpid();
-    rand_pool_add(pool, &curr_pid, sizeof(curr_pid), 0);
-    curr_uid = getuid();
-    rand_pool_add(pool, &curr_uid, sizeof(curr_uid), 0);
```
So uid/gid aren't used anymore? (I don't really care, just asking.)

kroeckx commented on crypto/rand/rand_unix.c:
```diff
@@ -256,5 +286,126 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+
+
+/*
+ * Get the current time with the highest possible accuracy
```
It's rather strange that the 3 things you try in order all return the same clock but with different resolution, while you speak about accuracy. But I do get what you mean anyway.

kroeckx commented:
Feel free to merge this with @levitte review.

mspncp commented:
Merged, thanks!

mspncp commented:
Wow! I hit the 22,000! :champagne: :champagne: :champagne: 

![22000](https://user-images.githubusercontent.com/11467854/38753022-62d43530-3f5d-11e8-8229-0d129dab52fd.png)
